### PR TITLE
ERR: Raise a simpler backtrace for missing key

### LINKED ISSUE #21557: KeyError error message could be less confusing
#### Code Sample, a copy-pastable example if possible

```python
In [2]: pd.Series(range(3)).loc[4]
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/home/nobackup/repo/pandas/pandas/core/indexing.py in _validate_key(self, key, axis)
   1810                 if not ax.contains(key):
-> 1811                     error()
   1812             except TypeError as e:

/home/nobackup/repo/pandas/pandas/core/indexing.py in error()
   1805                                .format(key=key,
-> 1806                                        axis=self.obj._get_axis_name(axis)))
   1807 

KeyError: 'the label [4] is not in the [index]'

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
<ipython-input-2-a40b2ecc79ee> in <module>()
----> 1 pd.Series(range(3)).loc[4]

/home/nobackup/repo/pandas/pandas/core/indexing.py in __getitem__(self, key)
   1497 
   1498             maybe_callable = com._apply_if_callable(key, self.obj)
-> 1499             return self._getitem_axis(maybe_callable, axis=axis)
   1500 
   1501     def _is_scalar_access(self, key):

/home/nobackup/repo/pandas/pandas/core/indexing.py in _getitem_axis(self, key, axis)
   1930 
   1931         # fall thru to straight lookup
-> 1932         self._validate_key(key, axis)
   1933         return self._get_label(key, axis=axis)
   1934 

/home/nobackup/repo/pandas/pandas/core/indexing.py in _validate_key(self, key, axis)
   1817                 raise
   1818             except:
-> 1819                 error()
   1820 
   1821     def _is_scalar_access(self, key):

/home/nobackup/repo/pandas/pandas/core/indexing.py in error()
   1804                 raise KeyError(u"the label [{key}] is not in the [{axis}]"
   1805                                .format(key=key,
-> 1806                                        axis=self.obj._get_axis_name(axis)))
   1807 
   1808             try:

KeyError: 'the label [4] is not in the [index]'
```

#### Problem description

The way in which we catch exceptions is uninteresting to the user. The sequence of two exceptions suggests to the user that there is a bug in the code, or unexpected exception, which is not the case; so this error message is confusing.

The fix is trivial.

#### Expected Output
```python
In [2]: pd.Series(range(3)).loc[4]
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-2-a40b2ecc79ee> in <module>()
----> 1 pd.Series(range(3)).loc[4]

/home/nobackup/repo/pandas/pandas/core/indexing.py in __getitem__(self, key)
   1497 
   1498             maybe_callable = com._apply_if_callable(key, self.obj)
-> 1499             return self._getitem_axis(maybe_callable, axis=axis)
   1500 
   1501     def _is_scalar_access(self, key):

/home/nobackup/repo/pandas/pandas/core/indexing.py in _getitem_axis(self, key, axis)
   1931 
   1932         # fall thru to straight lookup
-> 1933         self._validate_key(key, axis)
   1934         return self._get_label(key, axis=axis)
   1935 

/home/nobackup/repo/pandas/pandas/core/indexing.py in _validate_key(self, key, axis)
   1818 
   1819             if not ax.contains(key):
-> 1820                 error()
   1821 
   1822     def _is_scalar_access(self, key):

/home/nobackup/repo/pandas/pandas/core/indexing.py in error()
   1804                 raise KeyError(u"the label [{key}] is not in the [{axis}]"
   1805                                .format(key=key,
-> 1806                                        axis=self.obj._get_axis_name(axis)))
   1807 
   1808             try:

KeyError: 'the label [4] is not in the [index]'
```
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: b36b451a74bc16d7ea64c158a3cd33fbfb504068
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-6-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.UTF-8
LOCALE: it_IT.UTF-8

pandas: 0.24.0.dev0+132.gb36b451a7
pytest: 3.5.0
pip: 9.0.1
setuptools: 39.2.0
Cython: 0.25.2
numpy: 1.14.3
scipy: 0.19.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.5.6
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.4
blosc: None
bottleneck: 1.2.0dev
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.2.2.post1153+gff6786446
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: 1.3.0
xlsxwriter: 0.9.6
lxml: 4.1.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.2.1

</details>

### PR DESCRIPTION
**Pull request by toobaz**
- [x] closes #21557
- [x] tests passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -1818,6 +1816,9 @@ def error():
             except:
                 error()
```
Comment (topper-123) [inline pandas/core/indexing.py]: In my last few commits this except clause hid the cause of the underlying exception. Cant it just be removed altogether and let exceptions raise as they occur onstead of goinc through error()?

Comment (toobaz) [reply pandas/core/indexing.py]: > In my last few commits this except clause hid the cause of the underlying exception.

Well, in principle this is its exact goal ;-) Can you provide an example? I guess we should just specify the exceptions it catches.

File: doc/source/whatsnew/v0.23.2.txt [hunk 2]
```diff
@@ -61,6 +61,7 @@ Bug Fixes
 
 - Bug in :meth:`Index.get_indexer_non_unique` with categorical key (:issue:`21448`)
 - Bug in comparison operations for :class:`MultiIndex` where error was raised on equality / inequality comparison involving a MultiIndex with ``nlevels == 1`` (:issue:`21149`)
+- The backtrace from a ``KeyError`` is now shorted and clearer (:issue:`21557`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.2.txt]: can you be a little more clear here :> (this is when indexing!)

backtrace -> traceback

File: pandas/core/indexing.py [hunk 3]
```diff
@@ -1818,6 +1816,9 @@ def error():
...
 
+            if not ax.contains(key):
```
Comment (jreback) [inline pandas/core/indexing.py]: is it possible to add a test for this (maybe compare the exception message)?

File: doc/source/whatsnew/v0.23.2.txt (hunk 2 above)
Comment (jreback) [reply doc/source/whatsnew/v0.23.2.txt]: I think this should go in 0.24 as this might require a tricky backport (maybe) as the refactored code is in 0.24

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py (hunk 3 above)
Comment (toobaz) [reply pandas/core/indexing.py]: Unfortunately comparing the message is not enough (it didn't change - and is already tested). We would have to check the traceback, and I have no idea of how to do it - and _if_ we want to do it.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @toobaz

File: pandas/core/indexing.py (hunk 1 above)
Comment (toobaz) [reply pandas/core/indexing.py]: > Cant it just be removed altogether and let exceptions raise as they occur onstead of goinc through error()?

@topper-123 well, [maybe you're right](https://github.com/pandas-dev/pandas/pull/21594)!

**Pull request merged**